src/python_bindings/mod.rs

```diff
@@ -1,98 +1,165 @@
-use crate::regex::walk_fsm;
 use bincode::config;
+use bincode::{Decode, Encode};
 use pyo3::exceptions::PyValueError;
 use pyo3::prelude::*;
-use pyo3::types::PyDict;
+use pyo3::types::{PyAny, PyDict};
 use pyo3::wrap_pyfunction;
-use rustc_hash::{FxHashMap, FxHashSet};
+use rustc_hash::{FxHashMap as HashMap, FxHashSet as HashSet};
 use serde_json::Value;
+use tokenizers::FromPretrainedParameters;
+
+macro_rules! type_name {
+    ($obj:expr) => {
+        // Safety: obj is always initialized and tp_name is a C-string
+        unsafe { std::ffi::CStr::from_ptr((&*(&*$obj.as_ptr()).ob_type).tp_name) }
+    };
+}
 
-#[pyclass(name = "FSMInfo")]
-pub struct PyFSMInfo {
-    #[pyo3(get)]
-    initial: State,
-    #[pyo3(get)]
-    finals: FxHashSet<State>,
-    #[pyo3(get)]
-    transitions: FxHashMap<(State, TransitionKey), State>,
-    #[pyo3(get)]
-    alphabet_anything_value: TransitionKey,
-    #[pyo3(get)]
-    alphabet_symbol_mapping: FxHashMap<String, TransitionKey>,
+#[pyclass(name = "Guide", module = "outlines_core.fsm.outlines_core_rs")]
+#[derive(Clone, Debug, PartialEq, Encode, Decode)]
+pub struct PyGuide {
+    state: StateId,
+    index: PyIndex,
 }
 
-impl From<FSMInfo> for PyFSMInfo {
-    fn from(fsm_info: FSMInfo) -> Self {
-        PyFSMInfo {
-            initial: fsm_info.initial,
-            finals: fsm_info.finals,
-            transitions: fsm_info.transitions,
-            alphabet_anything_value: fsm_info.alphabet_anything_value,
-            alphabet_symbol_mapping: fsm_info.alphabet_symbol_mapping,
+#[pymethods]
+impl PyGuide {
+    #[new]
+    fn __new__(index: PyIndex) -> Self {
+        PyGuide {
+            state: index.get_initial_state(),
+            index,
         }
     }
-}
 
-// FIXME: could be costly, confirm if FSMInfo will actually be part of the interface
-impl From<&PyFSMInfo> for FSMInfo {
-    fn from(fsm_info: &PyFSMInfo) -> Self {
-        FSMInfo {
-            initial: fsm_info.initial,
-            finals: fsm_info.finals.clone(),
-            transitions: fsm_info.transitions.clone(),
-            alphabet_anything_value: fsm_info.alphabet_anything_value,
-            alphabet_symbol_mapping: fsm_info.alphabet_symbol_mapping.clone(),
+    fn get_start_tokens(&self) -> PyResult<Vec<TokenId>> {
+        self.index
+            .get_allowed_tokens(self.index.get_initial_state())
+            .ok_or(PyErr::new::<PyValueError, _>(
+                "Initial state is not in allowed tokens",
+            ))
+    }
+
+    fn read_next_token(&mut self, token_id: TokenId) -> PyResult<Vec<TokenId>> {
+        match self.index.get_next_state(self.state, token_id) {
+            Some(new_state) => {
+                self.state = new_state;
+                self.index
+                    .get_allowed_tokens(new_state)
+                    .ok_or(PyErr::new::<PyValueError, _>(format!(
+                        "No token ids found for the next state: {new_state}"
+                    )))
```

Comment:
N.B. update state after allowed tokens successfully retrieved, otherwise state will be corrupted, add tests.